lazy evaluation for git provided properties

frawa opened the pull request:
Resolves #874.

mathieucarbou commented:
Hello @frawa,
Thanks for starting such efforts.
Please make sure to use the code style present in the classes you modify and do not reformat ;-)
Happy new year!

mathieucarbou commented:
... also do not forget to squash your commits ;-)

hazendaz commented:
> ... also do not forget to squash your commits ;-)

will make sure to squash on merge.

mathieucarbou commented on .config/style.intellij-idea.eclipse.xml:
please do not forget to remove this file from the PR (eventually adding an entry in .ignore file)

frawa replied:
This was an attempt to enable applying IntelliJ code style without installing Idea.
Do you have a better approach using VS Code, or Eclipse, or command line?

mathieucarbou replied:
No... Vscode is a hell if you were used to intellij.
We are trying it in our team - we were all using intellij, sadly we cannot anymore.
Although I was able to find equivalent things for basic features, for more advanced ones and formatting, I didn't yet.

mathieucarbou replied:
Clang works fine for c++, I don't know if ther is an equivalent for java

mathieucarbou commented on license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/CopyrightRangeProvider.java:
```diff
@@ -76,50 +78,60 @@ public void close() {
-        copyrightYears = inceptionYear + "-" + copyrightEnd;
-      }
-      result.put(COPYRIGHT_YEARS_KEY, copyrightYears);
+    final var memo = new HashMap<String, Integer>();
+    final Function<String, Integer> memoGet = (k) -> {
+      return memo.computeIfAbsent(k, (key) -> {
+        try {
+          if (COPYRIGHT_LAST_YEAR_KEY.equals(key)) {
+            return gitLookup.getYearOfLastChange(document.getFile());
+          } else if (COPYRIGHT_CREATION_YEAR_KEY.equals(key)) {
+            return gitLookup.getYearOfCreation(document.getFile());
+          }
+          return null;
+        } catch (IOException | GitAPIException e) {
+          throw new RuntimeException(
+              "CopyrightRangeProvider error on file: " + document.getFile().getAbsolutePath() + ": " + e.getMessage(), e);
+        }
+      });
+    };
```
Even if I understand the memoizing effect of it, I think it can be written in a better form for code readability and future maintainance by introduce a static method:

```
    public static <T, R> Function<T, R> memoize(Function<T, R> function) {
        Map<T, R> cache = new HashMap<>();
        return input -> cache.computeIfAbsent(input, function);
    }
```

and then crating only 2 functions for the 2 cases.

frawa replied:
Unfortunately, Java's missing approach to checked exceptions for Lambdas prevents this slick solution.
I went with an atomic ref and a helper function to avoid code duplication.

mathieucarbou replied:
There is a way to hack around that using templates for exceptions but this is a hacky way lol ;-) but since not a public API would be acceptable. The atomic ref is fine also ;-)

mathieucarbou commented on license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:
```diff
@@ -86,35 +86,15 @@ public enum DateSource {
-        .map(String::toUpperCase)
-        .map(GitLookup.DateSource::valueOf)
-        .orElse(GitLookup.DateSource.AUTHOR);
+    final GitLookup.DateSource dateSource = Optional.ofNullable(props.get(COPYRIGHT_LAST_YEAR_SOURCE_KEY)).map(String::trim).map(String::toUpperCase).map(GitLookup.DateSource::valueOf).orElse(GitLookup.DateSource.AUTHOR);
```
can you please revert to the original format of the file and keep the existing code style of a source code ? this is to keep the diffs focused on the PR goal - no code reformat in the PRs. ;-) Thanks!

frawa replied:
Done

mathieucarbou commented on license-maven-plugin/src/main/java/com/mycila/maven/plugin/license/AbstractLicenseMojo.java:
```diff
@@ -669,7 +671,7 @@ private void executeForLicenseSet(final LicenseSet licenseSet, final Callback ca
         }
 
-        return perDoc;
+        return new AdjustedMap<>(perDoc, perDocAdjustments);
```
I like this hack, but it could be improved. The `perDoc` map could be this adjusted map (called differently - could be named `ForwardingMap`), where keys would be first fetched from the suppliers, otherwise from `file.name` otherwise from globalProperties maps.

there is

frawa replied:
`ForwardingMap` it is.

Including `file.name` with forwarding will be tricky, as it is part of the read only props passed to the adjusting call.
You'd end up with a deeply nested structure of forwarding maps, while accumulating resolver adjustments. (I tried that earlier and did not like it.)

mathieucarbou replied:
file.name could be hard coded or put in a map copied from the global props.

mathieucarbou reviewed (commented):
A few improvements and cleanup to do.
Thanks! This is a nice speed improvement!

mathieucarbou commented:
@frawa also do not forget to rebase;-)

mathieucarbou commented:
@frawa thanks! Will check the PR again today.
Can you please also squash all the commits into one ?

mathieucarbou commented:
@frawa : checked the PR. 
Introducing this idea of forwarding map now makes me think that we do not need all these supplier maps at all but only 1 new map implementation like a `LazyMap` would probably be cleaner. It allows to pass suppliers (memoized or not) with not touching the PropertiesProvider interface.

I will push a commit for that this afternoon to cleanup: using a custom map implementation will allow to leave a lot of the API in-place.

frawa commented:
> @frawa : checked the PR. Introducing this idea of forwarding map now makes me think that we do not need all these supplier maps at all but only 1 new map implementation like a `LazyMap` would probably be cleaner. It allows to pass suppliers (memoized or not) with not touching the PropertiesProvider interface.
> 
> I will push a commit for that this afternoon to cleanup: using a custom map implementation will allow to leave a lot of the API in-place.

That was my very first approach, see [this commit](https://github.com/frawa/license-maven-plugin/pull/1/commits/3751bdcb8e802f0f7bc3b947add6940466f54c7e#diff- eb856076d9adf427c53efe7f9f39b4da7280234b5320d08568f3c8b66c68191e), but I found it too much work to make `LazyMap` safe to use. 

The supplier map seems to convey the intent in a clearer way, IMHO.

I'll let you judge and take over to you liking.

mathieucarbou commented:
> > @frawa : checked the PR. Introducing this idea of forwarding map now makes me think that we do not need all these supplier maps at all but only 1 new map implementation like a `LazyMap` would probably be cleaner. It allows to pass suppliers (memoized or not) with not touching the PropertiesProvider interface.
> > I will push a commit for that this afternoon to cleanup: using a custom map implementation will allow to leave a lot of the API in-place.
> 
> That was my very first approach, see [this commit](https://github.com/frawa/license-maven-plugin/pull/1/commits/3751bdcb8e802f0f7bc3b947add6940466f54c7e#diff-eb856076d9adf427c53efe7f9f39b4da7280234b5320d08568f3c8b66c68191e), but I found it too much work to make `LazyMap` safe to use.
> 
> The supplier map seems to convey the intent in a clearer way, IMHO.
> 
> I'll let you judge and take over to you liking.

Yes more work to create this map, but less after. It has to be done anyway because this is not possible to modify the `PropertiesProvider` since this is public API to allow SPI extensions.

mathieucarbou commented:
@frawa : I have squashed your original commits into one to keep your work as-is, and I added a second commit to update this PR. 

I have introduced a `LazyMap`, supplier based, in order not to change the public API of `PropertiesProvider`, this lazy map being able to alternatively grab keys from a default supplier function (which is our global properties map).

Would you be able to test these changes to ensure they are still performant ?

I will wait for your feedback before merging.

Thanks!

Thanks!

frawa commented:
> @frawa : I have squashed your original commits into one to keep your work as-is, and I added a second commit to update this PR.
> 
> I have introduced a `LazyMap`, supplier based, in order not to change the public API of `PropertiesProvider`, this lazy map being able to alternatively grab keys from a default supplier function (which is our global properties map).
> 
> Only 3 files are updated following this change.
> 
> Would you be able to test these changes to ensure they are still performant ?
> 
> I will wait for your feedback before merging.
> 
> Thanks!

Nice.
Confirmed, the performance on the repo I am looking at is still improved by at least a factor 3.

hazendaz commented:
Thanks @frawa I've also seen same results on mybatis project.  Using a profiler, here are my results.

Using currently 4.6 release
<html><body>
<!--StartFragment-->
com.mycila:license-maven-plugin:4.6:format {execution: default} | 79.60 s
-- | --


<!--EndFragment-->
</body>
</html>

Using current master here (and a very few extra cleanup items I have - nothing impactful)

<html><body>
<!--StartFragment-->
om.mycila:license-maven-plugin:5.0.0-SNAPSHOT:format {execution: default} | 24.23 s
-- | --


<!--EndFragment-->
</body>
</html>

I repeated the formatting and touched a few files to make sure they are working as expected with same results.

In the overall project for mybatis I tested it on, its no longer the worst performing aspect.  Its not nearly as fast as general formatters for obvious reasons but huge improvement non the less.
